File: archive/sdk/javascript/src/modules/governance.ts
import { PawClient } from '../client';
import {
  Proposal,
  VoteParams,
  DepositParams,
  VoteOption,
  TxResult,
  GasOptions
} from '../types';

export class GovernanceModule {
  constructor(private client: PawClient) {}

  /**
   * Submit a text proposal
   */
  async submitTextProposal(
    proposer: string,
    title: string,
    description: string,
    initialDeposit: string,
    denom: string = 'upaw',
    options?: GasOptions
  ): Promise<TxResult> {
    const message = {
      typeUrl: '/cosmos.gov.v1beta1.MsgSubmitProposal',
      value: {
        content: {
          typeUrl: '/cosmos.gov.v1beta1.TextProposal',
          value: {
            title,
            description
          }
        },
        initialDeposit: [{ denom, amount: initialDeposit }],
        proposer
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(proposer, [message], options);
  }

  /**
   * Vote on a proposal
   */
  async vote(
    voter: string,
    params: VoteParams,
    options?: GasOptions
  ): Promise<TxResult> {
    const message = {
      typeUrl: '/cosmos.gov.v1beta1.MsgVote',
      value: {
        proposalId: params.proposalId,
        voter,
        option: params.option,
        metadata: params.metadata || ''
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(voter, [message], options);
  }

  /**
   * Deposit to a proposal
   */
  async deposit(
    depositor: string,
    params: DepositParams,
    options?: GasOptions
  ): Promise<TxResult> {
    const denom = params.denom || 'upaw';
    const message = {
      typeUrl: '/cosmos.gov.v1beta1.MsgDeposit',
      value: {
        proposalId: params.proposalId,
        depositor,
        amount: [{ denom, amount: params.amount }]
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(depositor, [message], options);
  }

  /**
   * Get all proposals
   */
  async getProposals(status?: number): Promise<Proposal[]> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const statusParam = status !== undefined ? `?proposal_status=${status}` : '';
      const response = await fetch(`${restEndpoint}/cosmos/gov/v1beta1/proposals${statusParam}`);
      if (!response.ok) {
        return [];
      }

      const data = await response.json();
      return data.proposals || [];
    } catch (error) {
      console.error('Error fetching proposals:', error);
      return [];
    }
  }

  /**
   * Get proposal by ID
   */
  async getProposal(proposalId: string): Promise<Proposal | null> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/gov/v1beta1/proposals/${proposalId}`);
      if (!response.ok) {
        return null;
      }

      const data = await response.json();
      return data.proposal || null;
    } catch (error) {
      console.error('Error fetching proposal:', error);
      return null;
    }
  }

  /**
   * Get votes for a proposal
   */
  async getVotes(proposalId: string): Promise<any[]> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/gov/v1beta1/proposals/${proposalId}/votes`);
      if (!response.ok) {
        return [];
      }

      const data = await response.json();
      return data.votes || [];
    } catch (error) {
      console.error('Error fetching votes:', error);
      return [];
    }
  }

  /**
   * Get vote for a specific voter
   */
  async getVote(proposalId: string, voter: string): Promise<any | null> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(
        `${restEndpoint}/cosmos/gov/v1beta1/proposals/${proposalId}/votes/${voter}`
      );
      if (!response.ok) {
        return null;
      }

      const data = await response.json();
      return data.vote || null;
    } catch (error) {
      console.error('Error fetching vote:', error);
      return null;
    }
  }

  /**
   * Get deposits for a proposal
   */
  async getDeposits(proposalId: string): Promise<any[]> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/gov/v1beta1/proposals/${proposalId}/deposits`);
      if (!response.ok) {
        return [];
      }

      const data = await response.json();
      return data.deposits || [];
    } catch (error) {
      console.error('Error fetching deposits:', error);
      return [];
    }
  }

  /**
   * Get tally for a proposal
   */
  async getTally(proposalId: string): Promise<any | null> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/gov/v1beta1/proposals/${proposalId}/tally`);
      if (!response.ok) {
        return null;
      }

      const data = await response.json();
      return data.tally || null;
    } catch (error) {
      console.error('Error fetching tally:', error);
      return null;
    }
  }

  /**
   * Get governance parameters
   */
  async getParams(paramsType: 'voting' | 'tallying' | 'deposit'): Promise<any | null> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/gov/v1beta1/params/${paramsType}`);
      if (!response.ok) {
        return null;
      }

      const data = await response.json();
      return data[`${paramsType}_params`] || null;
    } catch (error) {
      console.error('Error fetching params:', error);
      return null;
    }
  }

  /**
   * Check if proposal has passed quorum
   */
  async hasQuorum(proposalId: string): Promise<boolean> {
    const tally = await this.getTally(proposalId);
    const tallyParams = await this.getParams('tallying');

    if (!tally || !tallyParams) {
      return false;
    }

    const totalVotes = BigInt(tally.yes) + BigInt(tally.no) + BigInt(tally.abstain) + BigInt(tally.no_with_veto);
    const quorum = BigInt(tallyParams.quorum);

    // Simplified quorum check (actual implementation depends on total bonded tokens)
    return totalVotes > quorum;
  }

  /**
   * Get vote option name
   */
  getVoteOptionName(option: VoteOption): string {
    switch (option) {
      case VoteOption.YES:
        return 'Yes';
      case VoteOption.NO:
        return 'No';
      case VoteOption.ABSTAIN:
        return 'Abstain';
      case VoteOption.NO_WITH_VETO:
        return 'No with Veto';
      default:
        return 'Unspecified';
    }
  }
}

File: archive/sdk/python/paw/modules/governance.py
"""Governance module for proposal operations."""

from typing import List, Optional
from datetime import datetime
from dateutil import parser
from ..types import (
    Proposal,
    TallyResult,
    VoteParams,
    DepositParams,
    VoteOption,
    Coin,
    TxResult,
    GasOptions
)


class GovernanceModule:
    """Governance module for proposal operations."""

    def __init__(self, client):
        """Initialize governance module."""
        self.client = client

    async def submit_text_proposal(
        self,
        proposer: str,
        title: str,
        description: str,
        initial_deposit: str,
        denom: str = "upaw",
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Submit a text proposal."""
        message = {
            "@type": "/cosmos.gov.v1beta1.MsgSubmitProposal",
            "content": {
                "@type": "/cosmos.gov.v1beta1.TextProposal",
                "title": title,
                "description": description
            },
            "initial_deposit": [{"denom": denom, "amount": initial_deposit}],
            "proposer": proposer
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def vote(
        self,
        voter: str,
        params: VoteParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Vote on a proposal."""
        message = {
            "@type": "/cosmos.gov.v1beta1.MsgVote",
            "proposal_id": params.proposal_id,
            "voter": voter,
            "option": params.option,
            "metadata": params.metadata
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def deposit(
        self,
        depositor: str,
        params: DepositParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Deposit to a proposal."""
        message = {
            "@type": "/cosmos.gov.v1beta1.MsgDeposit",
            "proposal_id": params.proposal_id,
            "depositor": depositor,
            "amount": [{"denom": params.denom, "amount": params.amount}]
        }

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    async def get_proposals(self, status: Optional[int] = None) -> List[Proposal]:
        """Get all proposals."""
        try:
            url = "/cosmos/gov/v1beta1/proposals"
            if status is not None:
                url += f"?proposal_status={status}"

            data = await self.client.get(url)
            proposals = data.get("proposals", [])
            return [self._parse_proposal(p) for p in proposals]
        except Exception:
            return []

    async def get_proposal(self, proposal_id: str) -> Optional[Proposal]:
        """Get proposal by ID."""
        try:
            data = await self.client.get(f"/cosmos/gov/v1beta1/proposals/{proposal_id}")
            proposal_data = data.get("proposal")
            if proposal_data:
                return self._parse_proposal(proposal_data)
            return None
        except Exception:
            return None

    async def get_tally(self, proposal_id: str) -> Optional[TallyResult]:
        """Get tally for a proposal."""
        try:
            data = await self.client.get(f"/cosmos/gov/v1beta1/proposals/{proposal_id}/tally")
            tally_data = data.get("tally")
            if tally_data:
                return TallyResult(**tally_data)
            return None
        except Exception:
            return None

    def _parse_proposal(self, data: dict) -> Proposal:
        """Parse proposal data."""
        tally_data = data.get("final_tally_result", {})
        deposits = data.get("total_deposit", [])

        return Proposal(
            proposal_id=data.get("proposal_id", ""),
            status=data.get("status", 0),
            final_tally_result=TallyResult(
                yes=tally_data.get("yes", "0"),
                abstain=tally_data.get("abstain", "0"),
                no=tally_data.get("no", "0"),
                no_with_veto=tally_data.get("no_with_veto", "0")
            ),
            submit_time=parser.parse(data.get("submit_time", "")),
            deposit_end_time=parser.parse(data.get("deposit_end_time", "")),
            total_deposit=[Coin(denom=d["denom"], amount=d["amount"]) for d in deposits],
            voting_start_time=parser.parse(data.get("voting_start_time", "")),
            voting_end_time=parser.parse(data.get("voting_end_time", ""))
        )

    @staticmethod
    def get_vote_option_name(option: VoteOption) -> str:
        """Get vote option name."""
        return {
            VoteOption.YES: "Yes",
            VoteOption.NO: "No",
            VoteOption.ABSTAIN: "Abstain",
            VoteOption.NO_WITH_VETO: "No with Veto",
            VoteOption.UNSPECIFIED: "Unspecified"
        }.get(option, "Unknown")
